python/cudnn/gemm/ops/swiglu_moe.py

```
@@ -0,0 +1,325 @@
+        if tensor.device.type != "cuda":
+            raise ValueError(f"{prefix}: {name} must be a CUDA tensor, got {tensor.device}")
+        if not tensor.is_contiguous():
+            raise ValueError(f"{prefix}: {name} must be contiguous, got stride {tuple(tensor.stride())}")
+    if routed_x.dim() != 3 or routed_x.shape[0] != 1 or Wg.dim() != 3 or Wu.dim() != 3 or Wd.dim() != 3:
+        raise ValueError(
+            f"{prefix}: expected routed_x[1,S,H], Wg/Wu[E,I,H], Wd[E,H,I]; "
+            f"got {tuple(routed_x.shape)}, {tuple(Wg.shape)}, {tuple(Wu.shape)}, {tuple(Wd.shape)}"
+        )
+    if any(tensor.device != routed_x.device for _, tensor in tensors[1:]) or first_token_offset.device != routed_x.device:
+        raise ValueError(f"{prefix}: inputs and first_token_offset must be on the same CUDA device")
+    E, I, H = Wg.shape
```

Ruff (0.16.3)

[error] Ambiguous variable name: `I`

(E741)